marahgh/assignment3/wordSearch.cpp

```diff
@@ -0,0 +1,212 @@
+  if (std::find(words.begin(), words.end(), word) != words.end()) {
+    return true;
+  }
+  return false;
+}
+
+/*The grid class,
+contains lower-case letters from whichever alphabet the dictionary uses.*/
+class Grid {
+  char** matrix;
+  int rows;
+  int cols;
+  bool empty;
+
+ public:
+  /*creates an empty grid,
+  a grid with the given dimensions filled by ' ' charachters*/
+  Grid(int rows, int cols)
+      : matrix(nullptr), rows(rows), cols(cols), empty(true) {
+    matrix = new char*[rows];
+    for (int i = 0; i < rows; i++) {
+      matrix[i] = new char[cols];
+    }
+    vector<char> chars = vector<char>((rows * cols), ' ');
+    empty = true;
+    fillGrid(chars);
+  }
+
+  /*receives a vector of the charachters to to fill the grid with,
+  grid's rows number and grid's columns number.
+  creates the grid if the dimensions are OK , otherwise creats an empty grid*/
+  Grid(vector<char> chars, int rows, int cols)
+      : matrix(nullptr), rows(rows), cols(cols), empty(false) {
+    if (rows >= 0 && cols >= 0 && chars.size() == rows * cols) {
+      matrix = new char*[rows];
+      for (int i = 0; i < rows; i++) {
+        matrix[i] = new char[cols];
+      }
+      fillGrid(chars);
+    }
+  }
+  /*Fills the grid with the given values*/
+  bool fillGrid(vector<char> chars) {
+    empty = false;
+    if (chars.size() != rows * cols) {
+      return false;
+    }
+    for (int i = 0; i < rows; i++) {
+      for (int j = 0; j < cols; j++) {
+        matrix[i][j] = chars[(i * cols) + j];
+      }
+    }
+    return true;
+  }
+  /*Grid's destructor*/
+  ~Grid() {
+    if (matrix == nullptr) {
+      return;
+    }
+    for (int i = 0; i < rows; i++) {
+      delete[] matrix[i];
+    }
+    delete[] matrix;
+  }
+
+  /*returns the grid's rows number*/
+  int getRows() { return rows; }
+
+  /*returns the grid's columns number*/
+  int getCols() { return cols; }
+
+  /*return true if the grid is empty, and false otherwise*/
+  bool isEmpty() { return empty; }
+  /*returns the i'th row in the grid if the grid is not empty,and nullptr
+   * otherwise*/
+  char* operator[](int i) {
+    if (isEmpty()) {
+      return nullptr;
+    }
+    return matrix[i];
+  }
+};
+
+/*Dictionary class.
+Dictionary words are lowercase alphabet words*/
+class Dictionary {
+  vector<string> dictionary;
+
+  /****************************************************************************/
+  /* a private auxiliary method of isPrefix,
+  checks whether w2 is a prefix of w1*/
+  bool isWordPrefix(string w1, string w2) {
+    if (w1.size() < w2.size()) {
+      return false;
+    }
+    for (int i = 0; i < w2.size(); i++) {
+      if (w1[i] != w2[i]) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /*Removes hyphens("-") and converts letters to lower case(returns a legal
+  dictionary word)
+  if the word contains alphabet letters and hyphens, other wise returns an empty
+  word */
+  string fixWord(string word) {
+    string fixed = "";

```

Comment:
Constructing a new string is actually a bit inefficient. We could instead modify the one we were given (since it is a copy anyway).